In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("https://raw.githubusercontent.com/krishnaik06/Complete-Feature-Engineering/master/mercedesbenz.csv",
                 usecols=['X1', 'X2'])
df.head()

X1  X2
0  v  at
1  t  av
2  w   n
3  t   n
4  v   n

In [4]:
df["X1"].unique()

array(['v', 't', 'w', 'b', 'r', 'l', 's', 'aa', 'c', 'a', 'e', 'h', 'z',
       'j', 'o', 'u', 'p', 'n', 'i', 'y', 'd', 'f', 'm', 'k', 'g', 'q',
       'ab'], dtype=object)

In [5]:
df["X1"].nunique()

27

In [6]:
df['X1'].value_counts()

aa    833
s     598
b     592
l     590
v     408
r     251
i     203
a     143
c     121
o      82
w      52
z      46
u      37
e      33
m      32
t      31
h      29
y      23
f      23
j      22
n      19
k      17
p       9
g       6
d       3
q       3
ab      3
Name: X1, dtype: int64

In [7]:
#Number of different labels
counts = df['X1'].value_counts().sum()
counts

4209

In [8]:
df['X1'].value_counts()

aa    833
s     598
b     592
l     590
v     408
r     251
i     203
a     143
c     121
o      82
w      52
z      46
u      37
e      33
m      32
t      31
h      29
y      23
f      23
j      22
n      19
k      17
p       9
g       6
d       3
q       3
ab      3
Name: X1, dtype: int64

In [9]:
df['X1'].nunique()

27

In [10]:
df['X2'].value_counts()

as    1659
ae     496
ai     415
m      367
ak     265
r      153
n      137
s       94
f       87
e       81
aq      63
ay      54
a       47
t       29
k       25
i       25
b       21
ao      20
ag      19
z       19
d       18
ac      13
g       12
ap      11
y       11
x       10
aw       8
at       6
h        6
al       5
an       5
q        5
av       4
ah       4
p        4
au       3
am       1
j        1
af       1
l        1
aa       1
c        1
o        1
ar       1
Name: X2, dtype: int64

In [16]:
df.X1.value_counts().sort_values(ascending=False).head(10).index

#df.X1.value_counts().sort_values(ascending=False).head(10)#.index

Index(['aa', 's', 'b', 'l', 'v', 'r', 'i', 'a', 'c', 'o'], dtype='object')

In [17]:
#Checking the top 10 repeating columns
top_10_labels = [y for y in df.X1.value_counts().sort_values(ascending=False).head(10).index]
top_10_labels

['aa', 's', 'b', 'l', 'v', 'r', 'i', 'a', 'c', 'o']

In [18]:
#Top 10 columns in ascending order
df.X1.value_counts().sort_values(ascending=False).head(10)

aa    833
s     598
b     592
l     590
v     408
r     251
i     203
a     143
c     121
o      82
Name: X1, dtype: int64

**When you see in this output image you will notice that the aa label is repeating 833 times in the X1 column and going down this number is decreasing.**

**So we took the top 10 results from the top and we will convert these top 10 results into one-hot encoding and the labels not occurring in this top ten label list will turn into zero.**

In [19]:
df=pd.get_dummies(df['X1']).sample(10)
df 

a  aa  ab  b  c  d  e  f  g  h  ...  p  q  r  s  t  u  v  w  y  z
3547  0   0   0  0  0  0  0  0  0  0  ...  0  0  0  1  0  0  0  0  0  0
3463  0   0   0  0  0  0  0  0  0  0  ...  0  0  0  0  0  0  1  0  0  0
2131  0   1   0  0  0  0  0  0  0  0  ...  0  0  0  0  0  0  0  0  0  0
3734  0   0   0  0  0  0  0  0  0  0  ...  0  0  0  1  0  0  0  0  0  0
2635  0   0   0  0  0  0  0  0  0  0  ...  0  0  0  0  0  0  0  0  0  0
69    0   0   0  0  0  0  0  0  0  0  ...  0  0  0  1  0  0  0  0  0  0
1390  0   0   0  0  0  0  0  0  0  0  ...  0  0  0  0  0  0  1  0  0  0
2407  0   0   0  0  0  0  0  0  0  0  ...  0  0  0  0  0  0  0  0  0  0
3113  0   0   0  0  0  0  0  0  0  0  ...  0  0  0  0  0  0  0  0  0  0
2238  0   0   0  0  0  0  0  0  0  0  ...  0  0  0  0  0  0  0  0  0  0

[10 rows x 27 columns]

**Now, here we apply the one-hot encoding to all multi categorical variables. You can see how the top 10 labels are now converted into binary format.**

gen_dummy_features = pd.get_dummies(df['X2'],drop_first=True).sample(10)
#pd.concat([df[['X1', 'X2']], gen_dummy_features],axis=1)

In [29]:
pd.set_option("display.max_columns",None)
X = pd.concat([pd.get_dummies(df[col], drop_first = False,sparse=False) for col in df], axis=1, keys=df.columns)
X.head()


a aa    ab  b  c  d  e  f  g  h  i     j  k  l     m  n  o  p  q  r  s  \
      0  0  1  0  0  0  0  0  0  0  0  0  1  0  0  0  1  0  0  0  0  0  0  0   
3547  1  1  0  1  1  1  1  1  1  1  1  1  0  1  1  1  0  1  1  1  1  1  1  0   
3463  1  1  0  1  1  1  1  1  1  1  1  1  0  1  1  1  0  1  1  1  1  1  1  1   
2131  1  0  1  1  1  1  1  1  1  1  1  1  0  1  1  1  0  1  1  1  1  1  1  1   
3734  1  1  0  1  1  1  1  1  1  1  1  1  0  1  1  1  0  1  1  1  1  1  1  0   
2635  1  1  0  1  1  1  1  1  1  1  1  0  1  1  1  1  0  1  1  1  1  1  1  1   

         t  u  v     w  y  z  
      1  0  0  0  1  0  0  0  
3547  1  1  1  1  0  1  1  1  
3463  0  1  1  0  1  1  1  1  
2131  0  1  1  1  0  1  1  1  
3734  1  1  1  1  0  1  1  1  
2635  0  1  1  1  0  1  1  1